rust: Support custom `list` types

It would be very helpful for me to be able to use a custom type for a `list<T>` type, but this crate seems to assume that `list<T>` always maps directly to `Vec<T>`.

Ex:

```
wit_bindgen::generate!({
    inline: r#"
        package foo:bar;

        interface baz {
            type bytes = list<u8>;
            read: func() -> bytes;
        }

        world bzt {
            use baz.{bytes};
        }
    "#,
    with: {
        "foo:bar/baz/bytes": crate::Bytes,
    }
});

struct Bytes(Vec<u8>);
```

```
error[E0308]: mismatched types
  --> ~/Projects/wit-bindgen-custom-list-type/target/debug/build/wit-bindgen-rust-macro-4f3b8d44afc291c2/out/bzt0.rs:43:21
   |
12 |             pub fn read() -> crate::Bytes {
   |                              ------------ expected `Bytes` because of return type
...
43 |                     result5
   |                     ^^^^^^^ expected `Bytes`, found `Vec<_>`
   |
   = note: expected struct `Bytes`
              found struct `Vec<_>`
help: try wrapping the expression in `Bytes`
   |
43 |                     Bytes(result5)
   |                     ++++++       +
```

My usecase is a wasm component that runs Lua scripts where I want to be able pass wit-bindgen's generated structs directly to `mlua` using its support of `serde::{Serialize, Deserialize}`. This works great for everything except `list<u8>` which would most logically map to a lua string (which hold arbitrary binary data), but it is being pulled in as an array like any other `Vec<T>` would. The way to solve this on the `mlua` side is to wrap it in a newtype with a different `IntoLua` implementation. I would really like to avoid having to have a full set of parallel types and conversions for everything that uses my `type bytes = list<u8>;`.

I think a reasonable way to implement this would be to just require that the type implements `Into<Vec<T>>`, `From<Vec<T>>`, and `Deref<Target = T>` insert some `.into()`s if the list type has a type defined in the `with` section. Though I haven't figured out how well that would fit into how wit-bindgen currently works. Thoughts on if that is a good approach or if there's a better option? I'm happy to do the implementation if this is something you'd accept.

Here's a quick hack just to prove to myself that this could reasonably work: https://github.com/azdle/wit-bindgen/commit/4a5b8d11db0850586516d0d1b9039bed92f9e78c

--- a/crates/rust/Cargo.toml
+++ b/crates/rust/Cargo.toml
@@ -34,6 +34,7 @@ wit-bindgen = { path = '../guest-rust', features = ['async'] }
 test-helpers = { path = '../test-helpers' }
 # For use with the custom attributes test
 serde_json = "1"
+bytes = "1"
 
 [features]
 serde = ['dep:serde', 'wit-bindgen-core/serde']

--- a/crates/rust/src/bindgen.rs
+++ b/crates/rust/src/bindgen.rs
@@ -670,11 +670,15 @@ impl Bindgen for FunctionBindgen<'_, '_> {
                 let val = format!("vec{}", tmp);
                 let ptr = format!("ptr{}", tmp);
                 let len = format!("len{}", tmp);
+                let vec = self.r#gen.path_to_vec();
                 if realloc.is_none() {
                     self.push_str(&format!("let {} = {};\n", val, operands[0]));
                 } else {
                     let op0 = operands.pop().unwrap();
-                    self.push_str(&format!("let {} = ({}).into_boxed_slice();\n", val, op0));
+                    self.push_str(&format!(
+                        "let {} = <_ as Into<{vec}<_>>>::into({}).into_boxed_slice();\n",
+                        val, op0
+                    ));
                 }
                 self.push_str(&format!("let {} = {}.as_ptr().cast::<u8>();\n", ptr, val));
                 self.push_str(&format!("let {} = {}.len();\n", len, val));
@@ -691,7 +695,7 @@ impl Bindgen for FunctionBindgen<'_, '_> {
                 self.push_str(&format!("let {} = {};\n", len, operands[1]));
                 let vec = self.r#gen.path_to_vec();
                 let result = format!(
-                    "{vec}::from_raw_parts({}.cast(), {1}, {1})",
+                    "<_ as From<{vec}<_>>>::from({vec}::from_raw_parts({}.cast(), {1}, {1}))",
                     operands[0], len
                 );
                 results.push(result);

--- a/crates/rust/src/lib.rs
+++ b/crates/rust/src/lib.rs
@@ -432,7 +432,8 @@ impl RustWasm {
             }
 
             self.src.push_str("mod _rt {\n");
-            self.src.push_str("#![allow(dead_code, clippy::all)]\n");
+            self.src
+                .push_str("#![allow(dead_code, unused_imports, clippy::all)]\n");
             let mut emitted = IndexSet::new();
             while !self.rt_module.is_empty() {
                 for item in mem::take(&mut self.rt_module) {

--- a/crates/rust/tests/codegen.rs
+++ b/crates/rust/tests/codegen.rs
@@ -31,3 +31,140 @@ mod inline_and_path {
         generate_all,
     });
 }
+
+#[allow(unused)]
+mod newtyped_list {
+    use std::ops::Deref;
+
+    wit_bindgen::generate!({
+        inline: r#"
+        package test:newtyped-list;
+
+        interface byte {
+            type typed-list-of-byte = list<u8>;
+            type newtyped-list-of-byte = list<u8>;
+
+            record rec-of-lists {
+                l: list<u8>,
+                tl: typed-list-of-byte,
+                nl: newtyped-list-of-byte,
+            }
+
+            use-list-of-byte: func(l: list<u8>) -> list<u8>;
+            use-typed-list-of-byte: func(tl: typed-list-of-byte) -> typed-list-of-byte;
+            use-newtyped-list-of-byte: func(nl: newtyped-list-of-byte) -> newtyped-list-of-byte;
+            use-rec-of-lists: func(t: rec-of-lists) -> rec-of-lists;
+        }
+
+        interface noncopy-byte {
+            // this will be new-typed by a non-copy struct
+            type noncopy-byte = u8;
+
+            type newtyped-list-of-noncopy-byte = list<noncopy-byte>;
+            type typed-list-of-noncopy-byte = list<noncopy-byte>;
+
+            record rec-of-lists-of-noncopy-byte {
+                ntl: newtyped-list-of-noncopy-byte,
+                tl: typed-list-of-noncopy-byte,
+                l: list<noncopy-byte>,
+            }
+
+            use-newtyped-list-of-noncopy-byte: func(nl: newtyped-list-of-noncopy-byte) -> newtyped-list-of-noncopy-byte;
+            use-typed-list-of-noncopy-byte: func(tl: typed-list-of-noncopy-byte) -> typed-list-of-noncopy-byte;
+            use-list-of-noncopy-byte: func(l: list<noncopy-byte>) -> list<noncopy-byte>;
+            use-rec-of-lists-of-noncopy-byte: func(t: rec-of-lists-of-noncopy-byte) -> rec-of-lists-of-noncopy-byte;
+        }
+
+        world test {
+            import byte;
+            export byte;
+            import noncopy-byte;
+            export noncopy-byte;
+        }
+        "#,
+        with: {
+            "test:newtyped-list/byte/newtyped-list-of-byte": crate::newtyped_list::NewtypedListOfByte,
+            "test:newtyped-list/noncopy-byte/noncopy-byte": crate::newtyped_list::NoncopyByte,
+            "test:newtyped-list/noncopy-byte/newtyped-list-of-noncopy-byte": crate::newtyped_list::NewtypedListofNoncopyByte,
+        }
+    });
+
+    #[derive(Debug, Clone)]
+    pub struct NewtypedListOfByte(Vec<u8>);
+
+    impl From<Vec<u8>> for NewtypedListOfByte {
+        fn from(value: Vec<u8>) -> Self {
+            NewtypedListOfByte(value)
+        }
+    }
+
+    impl From<NewtypedListOfByte> for Vec<u8> {
+        fn from(value: NewtypedListOfByte) -> Self {
+            value.0
+        }
+    }
+
+    impl Deref for NewtypedListOfByte {
+        type Target = Vec<u8>;
+
+        fn deref(&self) -> &Self::Target {
+            &self.0
+        }
+    }
+
+    #[derive(Debug, Clone)]
+    pub struct NoncopyByte(u8);
+
+    #[derive(Debug, Clone)]
+    pub struct NewtypedListofNoncopyByte(Vec<NoncopyByte>);
+
+    impl From<Vec<NoncopyByte>> for NewtypedListofNoncopyByte {
+        fn from(value: Vec<NoncopyByte>) -> Self {
+            NewtypedListofNoncopyByte(value)
+        }
+    }
+
+    impl From<NewtypedListofNoncopyByte> for Vec<NoncopyByte> {
+        fn from(value: NewtypedListofNoncopyByte) -> Self {
+            value.0
+        }
+    }
+
+    impl Deref for NewtypedListofNoncopyByte {
+        type Target = Vec<NoncopyByte>;
+
+        fn deref(&self) -> &Self::Target {
+            &self.0
+        }
+    }
+}
+#[allow(unused)]
+mod retyped_list {
+    use std::ops::Deref;
+
+    wit_bindgen::generate!({
+        inline: r#"
+        package test:retyped-list;
+
+        interface bytes {
+            // this will be the `Bytes` type from the bytes crate
+            type retyped-list-as-bytes = list<u8>;
+
+            record rec-bytes {
+                rl: retyped-list-as-bytes,
+            }
+
+            use-retyped-list-as-bytes: func(ri: retyped-list-as-bytes) -> retyped-list-as-bytes;
+            use-rec-of-retyped-list-as-bytes: func(rl: retyped-list-as-bytes) -> retyped-list-as-bytes;
+        }
+
+        world test {
+            import bytes;
+            export bytes;
+        }
+        "#,
+        with: {
+            "test:retyped-list/bytes/retyped-list-as-bytes": bytes::Bytes,
+        }
+    });
+}

